src/entry_point.rs

```diff
@@ -7,27 +7,51 @@ use core::ptr;
 // application starts. _start is invoked directly by the Tock kernel; it
 // performs stack setup then calls rust_start. rust_start performs data
 // relocation and sets up the heap before calling the rustc-generated main.
-// rust_start and _start are tightly coupled: the order of rust_start's
-// parameters is designed to simplify _start's implementation.
+// rust_start and _start are tightly coupled.
 //
-// The memory layout is controlled by the linker script and these methods. These
-// are written for the following memory layout:
+// The memory layout is controlled by the linker script.
 //
-//     +----------------+ <- app_heap_break
-//     | Heap           |
-//     +----------------| <- heap_bottom
-//     | .data and .bss |
-//     +----------------+ <- stack_top
-//     | Stack          |
-//     | (grows down)   |
-//     +----------------+ <- mem_start
+// When the kernel gives control to us, we get r0-r3 values that is as follows.
+//
+//     +--------------+ <- (r2) mem.len()
+//     | Grant        |
+//     +--------------+
+//     | Unused       |
+//     +--------------+ <- (r3) app_heap_break
+//     | Heap         |         (hardcoded to mem_start + 3072 in
+//     +--------------|          Processs::create which could be lesser than
+//     | .bss         |          mem_start + stack + .data + .bss)
+//     +--------------|
+//     | .data        |
+//     +--------------+
+//     | Stack        |
+//     +--------------+ <- (r1) mem_start
+//
+//     +--------------+
+//     | Flash        |
+//     +--------------+ <- (r0) (.crt0_header/app_start)
+//     | Flash        |
+//     +--------------+
```

Comment:
In v3 added _protected region_ and  `.text`. I am not sure about adding `.data` here as it can create ambiguity (`.rodata` is actually in `.text` section)